src/range-to-string.js

```diff
@@ -0,0 +1,68 @@
+/* global Node */
+
+/**
+ * Return the next node after `node` in a tree order traversal of the document.
+ */
+function nextNode(node, skipChildren) {
+  if (!skipChildren && node.firstChild) {
+    return node.firstChild
+  } else if (node.nextSibling) {
+    return node.nextSibling
+  }
+  while (node.parentNode) {
+    node = node.parentNode
+    if (node.nextSibling) {
+      return node.nextSibling
+    }
+  }
+  /* istanbul ignore next */
+  return null
+}
+
+function firstNode(range) {
+  if (range.startContainer.nodeType === Node.ELEMENT_NODE) {
+    const node = range.startContainer.childNodes[range.startOffset]
+    return node || range.startContainer
```

Comment:
The invariant `firstNode(range) <= firstNodeAfter(range)` would still hold, which is the important condition for ensuring the `forEachNodeInRange` loop exits.

In practice, I think I've come across many times when browsers generate ranges with an offset that is at the end of the `startContainer`. I'm worried that in this scenario, we'll then count text that might have been part of a previous child of the `startContainer`. I cannot see any harm in using `|| nextNode(range.startContainer, true)`, can you?

In the case where this would result in something that comes after the end container, the range is collapsed.

Actually, thinking about this just now makes me realize that it's possible to create a range with a start that is past its end, and `collapsed` will be `true`. We should maybe check for `range.collapsed` before we call `forEachNodeInRange` and short circuit the whole operation, returning the empty string.

What do you think?